src/main/java/enmasse/kafka/bridge/SinkBridgeEndpoint.java

```diff
@@ -473,4 +377,258 @@ private ErrorCondition checkFilters(Object partition, Object offset) {
+				
+				if (!partitions.isEmpty()) {
+					
+					if (LOG.isDebugEnabled()) {
+						for (TopicPartition partition : partitions) {
+							LOG.debug("topic {} partition {}", partition.getTopic(), partition.getPartition());
+						}
+					}
+				
+					// Sender QoS unsettled (AT_LEAST_ONCE), need to commit offsets before partitions are revoked
+					
+					if (this.qos == ProtonQoS.AT_LEAST_ONCE) {
+						// commit all tracked offsets for partitions
+						SinkBridgeEndpoint.this.commitOffsets(true);
+					}
+				}
+			});
+			
+			this.consumer.partitionsAssignedHandler(partitions -> {
+				LOG.debug("Partitions assigned {}", partitions.size());
+				if (!partitions.isEmpty()) {
+					if (LOG.isDebugEnabled()) {
+						for (TopicPartition partition : partitions) {
+							LOG.debug("topic {} partition {}", partition.getTopic(), partition.getPartition());
+						}
+					}
+					partitionsAssigned();
+				} else {
+					ErrorCondition condition =
+							new ErrorCondition(Symbol.getSymbol(Bridge.AMQP_ERROR_NO_PARTITIONS),
+									"All partitions already have a receiver");
+					sendProtonError(condition);
+				}
+			});
+		});
+	}
+	
+	/**
+	 * Callback to process a kafka record
+	 * @param record The record
+	 */
+	private void handleKafkaRecord(KafkaConsumerRecord<K, V> record) {
+
+		LOG.debug("Processing key {} value {} partition {} offset {}", 
+				record.key(), record.value(), record.partition(), record.offset());
+
+		
+		LOG.debug("Fetched {} records [{}]", 1, this.qos);
+		switch (this.qos){
+		case AT_MOST_ONCE:
+			// Sender QoS settled (AT_MOST_ONCE) : commit immediately and start message sending
+			
+			// 1. when start of batch: immediate commit
+			if (startOfBatch()) {
+				LOG.debug("Start of batch in {} mode => commit()", this.qos);
+				this.consumer.commit(ar -> {
+					if (ar.failed()) {
+						{
+							LOG.error("Error committing ... {}", ar.cause().getMessage());
+							//sendProtonError(newError(Bridge.AMQP_ERROR_KAFKA_SYNC, "Error in commit", ar.cause()));
+						}
+					}
+				});
+			}
+			// 2. commit succeeded, so we can enqueue record for sending
+			LOG.debug("Received from Kafka partition {} [{}], key = {}, value = {}", record.partition(), record.offset(), record.key(), record.value());
+			
+			// 3. start message sending
+			protonSend(new KafkaMessage<K, V>(record.partition(), record.offset(), record.record()));
+
+			break;
+		case AT_LEAST_ONCE:
+			// Sender QoS unsettled (AT_LEAST_ONCE) : start message sending, wait end and commit
+			
+			LOG.debug("Received from Kafka partition {} [{}], key = {}, value = {}", record.partition(), record.offset(), record.key(), record.value());
+			
+			// 1. start message sending
+			protonSend(new KafkaMessage<K, V>(record.partition(), record.offset(), record.record()));
+			
+			if (endOfBatch()) {
+				LOG.debug("End of batch in {} mode => commitOffsets()", this.qos);
+				try {
+					// 2. commit all tracked offsets for partitions
+					commitOffsets(false);
```

Comment:
Let's imagine that we read 5 records from Kafka cluster and we have to send as AT LEAST ONCE ... in this case we add these 5 records to the offset tracker and send messages. Then in the handler when we are at last record (so endofbatch), the commitOffsets is called but .... no commit is executed because the previous 5 records aren't confirmed yet by the receiver (they are in the offset tracker but not delivered). Imagine that no other messages we'll be received from Kafka (because no one is sending). The handler won't be called anymore and so the commitOffsets won't be called; even if we receive the disposition from receiver and we call the offset tracker delivered, the commit to the Kafka cluster isn't called. In the previous implementation we have the commitOffsets code were executed in any case, trying to commit something if needed because we were in a loop (while) and the poll was executed but if no records were returned the commit part was executed in any case.